### ISSUE TITLE
Use libpd as a static library with CMake and MSVC

### ISSUE THREAD

**Issue by lukaspirkl** [libpd/libpd#315]
I added the libpd as a git submodule into vendor/libpd directory. Then I added following lines into my CMakeLists.txt:

```
set(CMAKE_THREAD_LIBS_INIT "${CMAKE_SOURCE_DIR}/vendor/pthreads/lib/x86/pthreadVC2.lib" CACHE INTERNAL "" FORCE)
set(PTHREADS_INCLUDE_DIR   "${CMAKE_SOURCE_DIR}/vendor/pthreads/include" CACHE INTERNAL "" FORCE)
add_subdirectory(${CMAKE_CURRENT_SOURCE_DIR}/vendor/libpd)
target_include_directories(${PROJECT_NAME} PUBLIC ${CMAKE_CURRENT_SOURCE_DIR}/vendor/libpd/libpd_wrapper)
target_include_directories(${PROJECT_NAME} PUBLIC ${CMAKE_CURRENT_SOURCE_DIR}/vendor/libpd/pure-data/src)
```

I added libpd_static into link libraries:

```
target_link_libraries(${PROJECT_NAME} glfw opengl32 libglew_static glm physfs-static spdlog yaml-cpp portaudio_static libpd_static)
```

For start, I added just the call to `libpd_init()` to the start of the application with the `#include "z_libpd.h"`. But I have the following linker error:

```
AudioDevice.obj : error LNK2019: unresolved external symbol __imp__libpd_init referenced in function "public: __thiscall AudioDevice::AudioDevice(unsigned int)" (??0AudioDevice@@QAE@I@Z)
```

I think there is some issue with the `EXTERN` definition in the `m_pd.h` file. Because when the lib file is compiled for the static usage the symbols don't contain the `__imp_` prefix. But when there is __declspec in the `EXTERN` it means that the linker will look for a symbol with the  `__imp_` prefix. I am not sure whether I understand the linking process correctly. So I don't know if the static linking is not supported on Windows or I am doing something wrong.

**Comment by Spacechild1** [comment]
Yup, I guess we need a definition like `PD_STATIC` in `m_pd.h` which prevents the addition of `__declspec` on both ends.

Usually, libraries do it the other way around by only exporting functions if explicitly told to do so. Example: https://github.com/xiph/opus/blob/d2f6805c3f36718a742c33c56eb945c2e10646ef/include/opus_defines.h#L66. We can't do that, however, because if would affect existing build systems.

**Comment by pierreguillot** [comment]
I manage to link Pd statically with CMake and MSVC without changes to m_pd.h but I use another CMakeLists.txt file. I'm not sure that support for the static library on MSVC has been tested with the CMakeLists.txt file of libpd. Perhaps, that should be tested before changing m_pd.h?

**Comment by Spacechild1** [comment]
Could it be that in your CMake file for building libpd you didn't add the `PD_INTERNAL` definition?

**Comment by pierreguillot** [comment]
It seems that PD_INTERNAL is set to 1 
https://github.com/pierreguillot/Camomile/blob/master/Dependencies/LibPdBuild/CMakeLists.txt

**Comment by Spacechild1** [comment]
hmmm... will do some checks later this week.

**Comment by danomatika** [comment]
Not that this helps, but the Makefile now has a STATIC variable to build a static lib.

**Comment by danomatika** [comment]
There is a PR related to this: #316

**Comment by Spacechild1** [comment]
#316 is not really related, I'm afraid. The definition of `EXTERN` must be fixed in `m_pd.h`.

**Comment by Spacechild1** [comment]
Actually, this has already been fixed on the Pd core side, as `EXTERN` can now be defined by the user: https://github.com/pure-data/pure-data/commit/aba5e1d5c09256189e853ae58efc4b5e1256ee43

Now libpd would have to explicitly define `EXTERN` as empty when building/using the static lib. I would suggest to have something like this on the top of `z_libpd.h`:
```
// ...

#ifdef __cplusplus
extern "C"
{
#endif

// Allow static linking on Windows
#ifdef LIBPD_STATIC
#define EXTERN extern
#endif

#include "m_pd.h"

// ...
```

`LIBPD_STATIC` would have to be defined both when building and using the static libpd library.

**Comment by danomatika** [comment]
@Spacechild1 Is this fixed upstream now?

**Comment by Spacechild1** [comment]
Only insofar as it is now possible to override `EXTERN`. My previous post already shows how this can be solved in libpd.
